design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+- If the above is rectified, Contour and the Operator still will conflict on setting status
+- CI failures that occur because of conflicts between Contour and the Operator are not detected unless we manually look at the scheduled Operator CI job
+
+## Goals
+- Improve implementation in Contour and Operator to mitigate current issues
+- Improve development processes and practices to ensure this conflicting state does not occur again
+
+## Non Goals
+- Drop support for features or projects
+
+## Remediation Steps
+
+- Set up alerting on Operator CI failures
+  - So we catch issues sooner.
+- Operator sets controller name in Contour ConfigMap
+  - To ensure Contour does not immediately exit on startup.
+- Operator stops setting any status on Gateway API resources
+  - To ensure Contour and Operator do not "fight" over status
+- Contour does not reject GatewayClasses if parameters ref is set
+  - At this point, we should have a workable intermediate solution for the moment and can cut a patch release
+  - Reconciliation will be duplicated but this should "work" for now
+  - User creates GatewayClass/Contour CRD with GatewayClass name
+  - Operator reconciles, does not set status
+  - User creates Gateway referencing GatewayClass
+  - Operator reconciles, creates a Contour
+  - Contour reconciles GatewayClass and Gateway, sets status, etc.
+- Operator deprecates Contour CRD GatewayClassRef field
+  - Currently the gateway reference has to be a GatewayClass name which is then validated to have the correct (hardcoded) controller string
+  - We should instead ensure the CRD has a field for setting the controller string the instance of Contour will watch for and reconcile GatewayClasses for
+  - The Operator should also set status to invalid on Contour objects that try to use a GatewayClass controller string that already exists, as it doesn't really make sense for multiple instances of Contour to watch for the same Controller string
+- Contour no longer requires Gateway namespace/name to be configured
+  - This is a planned deprecation anyway
+  - Will help with next step
+- Remove GatewayClass and Gateway controllers from Operator
+  - Planned deprecation anyways
+  - Users will no longer create a GatewayClass that references a Contour CRD to get a Contour that is configured for Gateway API.
+  - Instead, users will create a Contour CRD with a GatewayClass name and the operator will turn that into a Contour with the appropriate config.
+  - Operator will need to expand on Contour CRD controller, ensure Contour deployment resources are created when GatewayClass ref is set on Contour object
+- Move Operator into main Contour repository (long-term/optional)
+  - This option is a longer term project that would help us sort out process issues between Contour and Operator development.
+  - We would move the Operator codebase into the Contour one and deprecate/remove the Operator specific repo.
+  - Pros
```

Comment:
I also propose that we don't need to release the operator 1:1 with Contour 100%.  The versions should be 1:1 and match accordingly, but we can skip an Operator release if needed on account of breaking changes in Gateway API or something else critical.  It's better to skip a release than rush a release.